README
This package has the Vyatta configuration and operational templates and scripts for clustering.


debian/changelog
vyatta-cluster (1.3-1) unstable; urgency=medium

  * New LTS branch.

 -- Daniil Baturin <daniil@vyos.io>  Mon, 04 Jan 2021 17:13:11 +0200

vyatta-cluster (0.11.25+vyos2+current1) unstable; urgency=medium

  [ Thomas Jepp ]
  * Fix build depends.

  [ Kim Hagen ]

 -- Kim Hagen <kim.sidney@gmail.com>  Sun, 24 Jan 2016 15:05:30 -0500

vyatta-cluster (0.11.25+vyos2+lithium4) unstable; urgency=low

  * vyatta-cluster: update dh_gencontrol with new development build flag

 -- Alex Harpin <development@landsofshadow.co.uk>  Tue, 16 Jun 2015 20:40:44 +0100
 
vyatta-cluster (0.11.25+vyos2+lithium3) unstable; urgency=low

  [ Alex Harpin ]
  * vyatta-cluster: add op mode restart command for clustering
  * vyatta-cluster: formatting changes for style consistency
  * vyatta-cluster: wait for link up on clustering for up to 10 seconds

 -- Alex Harpin <development@landsofshadow.co.uk>  Mon, 09 Mar 2015 08:49:28 +0000
 
vyatta-cluster (0.11.25+vyos2+lithium2) unstable; urgency=low

  * Update maintainer address

 -- Alex Harpin <development@landsofshadow.co.uk>  Fri, 26 Dec 2014 09:28:56 +0000
 
vyatta-cluster (0.11.25+vyos2+lithium1) unstable; urgency=low

  * New branch

 -- Alex Harpin <development@landsofshadow.co.uk>  Sun, 14 Dec 2014 12:32:28 +0000
 
vyatta-cluster (0.11.25+vyos1+helium1) unstable; urgency=low

  * New branch

 -- Daniil Baturin <daniil@baturin.org>  Sat, 15 Feb 2014 16:57:02 +0100

vyatta-cluster (0.11.25+hydrogen1) unstable; urgency=low

  * New branch

 -- Daniil Baturin <daniil@baturin.org>  Sun, 17 Nov 2013 00:18:57 +0100

vyatta-cluster (0.11.25+daisy1) unstable; urgency=low

  * create daisy branch

 -- John Southworth <john.southworth@vyatta.com>  Sat, 13 Oct 2012 13:30:48 -0700

vyatta-cluster (0.11.25) unstable; urgency=low

  * new branch

 -- John Southworth <john.southworth@vyatta.com>  Fri, 12 Oct 2012 19:47:04 -0700

vyatta-cluster (0.11.24) unstable; urgency=low

  * new branch

 -- Deepti Kulkarni <deepti@vyatta.com>  Sat, 03 Mar 2012 02:25:43 -0800

vyatta-cluster (0.11.23) unstable; urgency=low

  * fix 3238 - Removed auto-failback configuration warning.

 -- Deepti Kulkarni <deepti@vyatta.com>  Tue, 28 Feb 2012 03:37:17 -0800

vyatta-cluster (0.11.22) unstable; urgency=low

  * Added missing validation for force cluster standby command.

 -- Deepti Kulkarni <deepti@vyatta.com>  Wed, 27 Jul 2011 09:19:26 -0700

vyatta-cluster (0.11.21) unstable; urgency=low

  * Added op-mode commands

 -- Deepti Kulkarni <deepti@vyatta.com>  Wed, 27 Jul 2011 07:52:34 -0700

vyatta-cluster (0.11.20) unstable; urgency=low

  * new branch

 -- Deepti Kulkarni <deepti@vyatta.com>  Thu, 07 Jul 2011 20:56:15 -0700

vyatta-cluster (0.11.19) unstable; urgency=low

  * fix for 6107. added val_help for correct u32 range.

 -- Michael Larson <mike@vyatta.com>  Fri, 13 May 2011 15:25:09 -0700

vyatta-cluster (0.11.18) unstable; urgency=low

  * add interface to help text

 -- An-Cheng Huang <ancheng@vyatta.com>  Wed, 02 Mar 2011 09:41:22 -0800

vyatta-cluster (0.11.17) unstable; urgency=low

  * new branch

 -- An-Cheng Huang <ancheng@vyatta.com>  Tue, 28 Dec 2010 13:48:14 -0800

vyatta-cluster (0.11.16) unstable; urgency=low

  * UNRELEASED

 -- An-Cheng Huang <ancheng@vyatta.com>  Thu, 02 Sep 2010 18:28:42 -0700

vyatta-cluster (0.11.15) unstable; urgency=low

  * update help text to use val_help

 -- An-Cheng Huang <ancheng@vyatta.com>  Tue, 17 Aug 2010 15:31:16 -0700

vyatta-cluster (0.11.14) unstable; urgency=low

  * UNRELEASED

 -- An-Cheng Huang <ancheng@vyatta.com>  Thu, 22 Jul 2010 17:23:56 -0700

vyatta-cluster (0.11.13) unstable; urgency=low

  * remove verb usage from the begining of help strings

 -- Mohit Mehta <mohit.mehta@vyatta.com>  Wed, 21 Jul 2010 18:22:30 -0700

vyatta-cluster (0.11.12) unstable; urgency=low

  * hide start-stop messages from configured services

 -- Mohit Mehta <mohit.mehta@vyatta.com>  Mon, 12 Jul 2010 14:41:11 -0700

vyatta-cluster (0.11.11) unstable; urgency=low

  * use more clear error message

 -- Mohit Mehta <mohit.mehta@vyatta.com>  Fri, 11 Jun 2010 15:26:27 -0700

vyatta-cluster (0.11.10) unstable; urgency=low

  * move cluster conntrack-sync script to vyatta-conntrack-sync package

 -- Mohit Mehta <mohit.mehta@vyatta.com>  Thu, 03 Jun 2010 12:02:11 -0700

vyatta-cluster (0.11.9) unstable; urgency=low

  * fix for bug 5134

 -- An-Cheng Huang <ancheng@vyatta.com>  Wed, 02 Jun 2010 09:15:10 -0700

vyatta-cluster (0.11.8) unstable; urgency=low

  * make clustering work with conntrack-sync

 -- Mohit Mehta <mohit.mehta@vyatta.com>  Fri, 28 May 2010 16:34:19 -0700

vyatta-cluster (0.11.7) unstable; urgency=low

  * UNRELEASED

 -- An-Cheng Huang <ancheng@vyatta.com>  Wed, 17 Feb 2010 16:13:58 -0800

vyatta-cluster (0.11.6) unstable; urgency=low

  * added required keyword to help text.

 -- Michael Larson <slioch@slioch.vyatta.com>  Mon, 30 Nov 2009 15:23:54 -0800

vyatta-cluster (0.11.5) unstable; urgency=low

  * dependency update

 -- Michael Larson <slioch@slioch.vyatta.com>  Fri, 13 Nov 2009 14:06:25 -0800

vyatta-cluster (0.11.4) unstable; urgency=low

  * updated priority node.def for project

 -- slioch <slioch@eng-140.vyatta.com>  Wed, 21 Oct 2009 14:55:12 -0700

vyatta-cluster (0.11.3) unstable; urgency=low

  * Fix 4623: Removing IPSEC VPN config without removing cluster ipsec
    config drops all interfaces.

 -- Stig Thormodsrud <stig@vyatta.com>  Fri, 10 Jul 2009 14:13:32 -0700

vyatta-cluster (0.11.2) unstable; urgency=low

  * UNRELEASED

 -- An-Cheng Huang <ancheng@vyatta.com>  Fri, 29 May 2009 18:35:46 -0700

vyatta-cluster (0.11.1) unstable; urgency=low

  [ An-Cheng Huang ]

scripts/VyattaClusterConfig.pm
package VyattaClusterConfig;

use strict;
use lib "/opt/vyatta/share/perl5/";
use VyattaConfig;

my $DEFAULT_INITDEAD = 30000;
my $DEFAULT_DEADPING = 30000;
my $MIN_DEAD = 300;
my $MIN_KEEP = 100;
my $DEFAULT_LOG_FACILITY = 'daemon';
my $SERVICE_DIR = "/etc/init.d";
my $RESOURCE_SCRIPT_DIR = "/etc/ha.d/resource.d";

# for heartbeat
my $DEFAULT_MCAST_GROUP = '239.251.252.253';
my $DEFAULT_UDP_PORT = '694';
my $DEFAULT_TTL = '1';

my $HA_WATCHLINK_ID = 'ha';

my %fields = (
  _interface        => undef,
  _mcast_grp        => undef,
  _pre_shared       => undef,
  _keepalive_itvl   => undef,
  _dead_itvl        => undef,
  _groups           => {},
  _is_empty         => 1,
);

sub new {
  my $that = shift;
  my $class = ref ($that) || $that;
  my $self = {
    %fields,
  };

  bless $self, $class;
  return $self;
}

sub setup {
  my ( $self, $level ) = @_;
  my $config = new VyattaConfig;

  $config->setLevel("$level");
  my @nodes = $config->listNodes();
  if (scalar(@nodes) <= 0) {
    $self->{_is_empty} = 1;
    return 0;
  } else {
    $self->{_is_empty} = 0;
  }

  my @tmp = $config->returnValues("interface");
  $self->{_interface} = [ @tmp ];
  $self->{_mcast_grp} = $config->returnValue("mcast-group");
  $self->{_pre_shared} = $config->returnValue("pre-shared-secret");
  $self->{_keepalive_itvl} = $config->returnValue("keepalive-interval");
  $self->{_dead_itvl} = $config->returnValue("dead-interval");

  $config->setLevel("$level group");
  my @groups = $config->listNodes();
  my $group;
  for $group (@groups) {
    my $hashref = {};
    $config->setLevel("$level group $group");
    $hashref->{_primary} = $config->returnValue("primary");
    @tmp = $config->returnValues("secondary");
    $hashref->{_secondary} = [ @tmp ];
    $hashref->{_auto_failback} = $config->returnValue("auto-failback");
    @tmp = $config->returnValues("monitor");
    $hashref->{_monitor} = [ @tmp ];
    @tmp = $config->returnValues("service");
    $hashref->{_service} = [ @tmp ];
    $self->{_groups}->{$group} = $hashref;
  }

  return 0;
}

sub setupOrig {
  my ( $self, $level ) = @_;
  my $config = new VyattaConfig;

  $config->setLevel("$level");
  my @nodes = $config->listOrigNodes();
  if (scalar(@nodes) <= 0) {
    $self->{_is_empty} = 1;
    return 0;
  } else {
    $self->{_is_empty} = 0;
  }

  my @tmp = $config->returnOrigValues("interface");
  $self->{_interface} = [ @tmp ];
  $self->{_mcast_grp} = $config->returnOrigValue("mcast-group");
  $self->{_pre_shared} = $config->returnOrigValue("pre-shared-secret");
  $self->{_keepalive_itvl} = $config->returnOrigValue("keepalive-interval");
  $self->{_dead_itvl} = $config->returnOrigValue("dead-interval");

  $config->setLevel("$level group");
  my @groups = $config->listOrigNodes();
  my $group;
  for $group (@groups) {
    my $hashref = {};
    $config->setLevel("$level group $group");
    $hashref->{_primary} = $config->returnOrigValue("primary");
    @tmp = $config->returnOrigValues("secondary");
    $hashref->{_secondary} = [ @tmp ];
    $hashref->{_auto_failback} = $config->returnOrigValue("auto-failback");
    @tmp = $config->returnOrigValues("monitor");
    $hashref->{_monitor} = [ @tmp ];
    @tmp = $config->returnOrigValues("service");
    $hashref->{_service} = [ @tmp ];
    $self->{_groups}->{$group} = $hashref;
  }

  return 0;
}

sub primaryNode {
  my ($self) = @_;
  my @groups = keys %{$self->{_groups}};
  my $hashref = $self->{_groups}->{$groups[0]};
  return $hashref->{_primary};
}

sub secondaryNode {
  my ($self) = @_;
  my @groups = keys %{$self->{_groups}};
  my $hashref = $self->{_groups}->{$groups[0]};
  return ${$hashref->{_secondary}}[0];
}

sub monitorNodes {
  my ($self) = @_;
  my @groups = keys %{$self->{_groups}};
  my $hashref = $self->{_groups}->{$groups[0]};
  return @{$hashref->{_monitor}};
}

sub serviceStr {
  my ($self) = @_;
  my @groups = keys %{$self->{_groups}};
  my $hashref = $self->{_groups}->{$groups[0]};
  return (join " ", @{$hashref->{_service}});
}

sub isEmpty {
  my ($self) = @_;
  return $self->{_is_empty};
}

sub authkeys {
  my ($self) = @_;
  my $key = $self->{_pre_shared};
  return (undef, "pre-shared secret not defined") if (!defined($key));
  my $str =<<EOS;
auth 1
1 sha1 $key
EOS
  return ($str, undef);
}

sub check_interfaces {
  my ($exist_only, @interfaces) = @_;
  foreach (@interfaces) {
    system("ip addr show $_ >& /dev/null");
    if ($? >> 8) {
      return "interface $_ does not exist";
    }
    next if ($exist_only);

    my $link = `ip link show $_ | grep $_`;
    if (($link =~ /NO-CARRIER/) || !($link =~ /,UP/)) {
      return "interface $_ is not connected";
    }
    system("ip addr show dev $_ |grep 'inet ' |grep -q 'scope global'");
    if ($? >> 8) {
      return "interface $_ is not configured";
    }
  }
  return undef;
}

sub ha_cf {
  my ($self) = @_;
  my @groups = keys %{$self->{_groups}};
  return (undef, "no resource group defined") if ($#groups < 0);
  return (undef, "using multiple resource groups is not supported yet")
    if ($#groups > 0);

  my $ierr = check_interfaces(0, @{$self->{_interface}});
  if (defined($ierr)) {
    return (undef, $ierr);
  }
  my $interfaces = '';
  foreach my $intf (@{$self->{_interface}}) {
    $interfaces .= "mcast $intf ";
    $interfaces .= ((defined($self->{_mcast_grp}))
                    ?  "$self->{_mcast_grp} " : "$DEFAULT_MCAST_GROUP ");
    $interfaces .= "$DEFAULT_UDP_PORT $DEFAULT_TTL 0\n";
  }

  my $kitvl = $self->{_keepalive_itvl};
  my $ditvl = $self->{_dead_itvl};

  my $hashref = $self->{_groups}->{$groups[0]};
  my $primary = $hashref->{_primary};
  my @secondaries =  @{$hashref->{_secondary}};
  my $pings = join " ", @{$hashref->{_monitor}};
  my $auto_failback = ($hashref->{_auto_failback} eq "true") ?
                      "on" : "off";
  my $my_name = `uname -n`;
  chomp $my_name;

  return (undef, "heartbeat interface(s) not defined") if ($interfaces eq "");
  return (undef, "keepalive interval not defined") if (!defined($kitvl));
  return (undef, "dead interval not defined") if (!defined($ditvl));
  return (undef, "cluster primary system not defined")
    if (!defined($primary));
  return (undef, "cluster secondary node(s) not defined")
    if ($#secondaries < 0);
  return (undef, "using multiple secondary nodes is not supported yet")
    if ($#secondaries > 0);
  return (undef,
          "dead interval must be at least $MIN_DEAD milliseconds")
    if ($ditvl < $MIN_DEAD);
  return (undef,
          "keepalive interval must be at least $MIN_KEEP milliseconds")
    if ($kitvl < $MIN_KEEP);
  return (undef,
          "dead interval must be more than twice the keepalive interval")
    if ($ditvl <= (2 * $kitvl));
  return (undef,
          "dead interval must be smaller than $DEFAULT_INITDEAD milliseconds")
    if ($ditvl >= $DEFAULT_INITDEAD);
  return (undef,
          "the current node '$my_name' is not defined in the configuration")
    if (($my_name ne $primary) && ($my_name ne $secondaries[0]));
 
  my $monitor_str = "";
  if ($pings ne "") {
    $monitor_str = "\nping $pings\n"
                   . "respawn hacluster /usr/lib/heartbeat/ipfail";
  }

  my $wtime = int($kitvl * 2);
  if ($wtime > $ditvl) {
    $wtime = $ditvl;
  }

  # convert to seconds (HA calls "sleep" with this)
  $ditvl /= 1000;

  my $str =<<EOS;
keepalive ${kitvl}ms
deadtime ${ditvl}
warntime ${wtime}ms
initdead ${DEFAULT_INITDEAD}ms
deadping ${DEFAULT_DEADPING}ms
logfacility $DEFAULT_LOG_FACILITY
${interfaces}auto_failback $auto_failback
node $primary $secondaries[0]$monitor_str
EOS

  return ($str, undef);
}

sub isValidIPSpec {
  my $str = shift;
  my @comps = split /\//, $str;
  return 0 if ($#comps < 1);
  return 0 if ($#comps > 3);
  return 0 if (!isValidIPv4($comps[0]));
  # check optional prefix len
  if (defined($comps[1])) {
    return 0 if (!($comps[1] =~ m/^\d+$/));
    return 0 if (($comps[1] > 32) || ($comps[1] < 0));
  }
  # check optional interface
  if (defined($comps[2])) {
    return 0 if (defined(check_interfaces(1, $comps[2])));
  }
  # check optional broadcast addr
  if (defined($comps[3])) {
    return 0 if (!isValidIPv4($comps[3]));
  }
  return 1;
}

sub isValidIPv4 {
  my $str = shift;
  return 0 if (!($str =~ m/^(\d+)\.(\d+)\.(\d+)\.(\d+)$/));
  my ($a1, $a2, $a3, $a4) = ($1, $2, $3, $4);
  return 1 if (($a1 >= 0 && $a1 < 256) && ($a2 >= 0 && $a2 < 256)
               && ($a3 >= 0 && $a3 < 256) && ($a4 >= 0 && $a4 < 256));
  return 0;
}

my @service_list = ();
sub isValidService {
  my $service = shift;
  if ($service =~ /^([^:]+)::/) {
    my $script = $1;
    return 0 if (! -e "$RESOURCE_SCRIPT_DIR/$script");
    return 1;
  }
  if (scalar(@service_list) == 0) {
    opendir(SDIR, "$SERVICE_DIR")
      or (print STDERR "Error: can't open $SERVICE_DIR" && return 0);
    @service_list = grep !/^\./, readdir SDIR;
  }
  return 1 if (grep /^$service$/, @service_list);
  return 0;
}

sub haresources {
  my ($self) = @_;
  my @groups = keys %{$self->{_groups}};
  return (undef, "no resource group defined") if ($#groups < 0);
  return (undef, "using multiple resource groups is not supported yet")
    if ($#groups > 0);
  
  my $hashref = $self->{_groups}->{$groups[0]};
  my $primary = $hashref->{_primary};

  my @init_services = ();
  my @ip_addresses = ();
  foreach (@{$hashref->{_service}}) {
    if (!isValidIPSpec($_)) {
      if (isValidService($_)) {
        push @init_services, $_;
      } else {
        return (undef, "\"$_\" is not a valid IP address "
                       . "(with subnet mask length) or service name");
      }
    } else {
      push @ip_addresses, "IPaddr2-vyatta::$_";
    }
  }
  # this forces all ip addresses to be before all services, which may not
  # be the desirable behavior in all cases.
  my $ip_str = join " ", @ip_addresses;
  my $serv_str = join " ", @init_services;
  my $services = join " ", ($ip_str, $serv_str);
  return (undef, "cluster primary system not defined") if (!defined($primary));
  return (undef, "cluster service(s) not defined") if ($services eq "");

  my $str =<<EOS;
$primary $services
EOS
  
  return ($str, undef, @init_services);
}

sub print_str {
  my ($self) = @_;
  my $str = "cluster";
  $str .= "\n  interface " . (join ",", @{$self->{_interface}});
  $str .= "\n  pre-shared-secret $self->{_pre_shared}";
  $str .= "\n  keepalive-interval $self->{_keepalive_itvl}";
  $str .= "\n  dead-interval $self->{_dead_itvl}";
  my $group;
  foreach $group (keys %{$self->{_groups}}) {
    $str .= "\n  group $group";
    my $hashref = $self->{_groups}->{$group};
    $str .= "\n    primary $hashref->{_primary}";
    $str .= "\n    secondary " . (join ",", @{$hashref->{_secondary}});
    $str .= "\n    monitor " . (join ",", @{$hashref->{_monitor}});
    $str .= "\n    service " . (join ",", @{$hashref->{_service}});
  }

  return $str;
}

1;



lib/Vyatta/Cluster/Config.pm
package Vyatta::Cluster::Config;

use strict;
use lib "/opt/vyatta/share/perl5";
use Vyatta::Config;

my $DEFAULT_INITDEAD = 30000;
my $MIN_DEAD = 300;
my $MIN_KEEP = 100;
my $DEFAULT_LOG_FACILITY = 'daemon';
my $SERVICE_DIR = "/etc/init.d";
my $RESOURCE_SCRIPT_DIR = "/etc/ha.d/resource.d";

# for heartbeat
my $DEFAULT_MCAST_GROUP = '239.251.252.253';
my $DEFAULT_UDP_PORT = '694';
my $DEFAULT_TTL = '1';

my $HA_WATCHLINK_ID = 'ha';

my %fields = (
    _interface          => undef,
    _mcast_grp          => undef,
    _pre_shared         => undef,
    _keepalive_itvl     => undef,
    _dead_itvl          => undef,
    _monitor_dead_itvl  => undef,
    _groups             => {},
    _is_empty           => 1,
);

sub new {
    my $that = shift;
    my $class = ref($that) || $that;
    my $self = {%fields,};

    bless $self, $class;
    return $self;
}

sub setup {
    my ($self, $level) = @_;
    my $config = new Vyatta::Config;

    $config->setLevel("$level");
    my @nodes = $config->listNodes();
    if (scalar(@nodes) <= 0) {
        $self->{_is_empty} = 1;
        return 0;
    } else {
        $self->{_is_empty} = 0;
    }

    my @tmp = $config->returnValues("interface");
    $self->{_interface} = [@tmp];
    $self->{_mcast_grp} = $config->returnValue("mcast-group");
    $self->{_pre_shared} = $config->returnValue("pre-shared-secret");
    $self->{_keepalive_itvl} = $config->returnValue("keepalive-interval");
    $self->{_dead_itvl} = $config->returnValue("dead-interval");
    $self->{_monitor_dead_itvl} = $config->returnValue("monitor-dead-interval");

    $config->setLevel("$level group");
    my @groups = $config->listNodes();
    my $group;
    for $group (@groups) {
        my $hashref = {};
        $config->setLevel("$level group $group");
        $hashref->{_primary} = $config->returnValue("primary");
        @tmp = $config->returnValues("secondary");
        $hashref->{_secondary} = [@tmp];
        $hashref->{_auto_failback} = $config->returnValue("auto-failback");
        @tmp = $config->returnValues("monitor");
        $hashref->{_monitor} = [@tmp];
        @tmp = $config->returnValues("service");
        $hashref->{_service} = [@tmp];
        $self->{_groups}->{$group} = $hashref;
    }

    return 0;
}

sub setupOrig {
    my ($self, $level) = @_;
    my $config = new Vyatta::Config;

    $config->setLevel("$level");
    my @nodes = $config->listOrigNodes();
    if (scalar(@nodes) <= 0) {
        $self->{_is_empty} = 1;
        return 0;
    } else {
        $self->{_is_empty} = 0;
    }

    my @tmp = $config->returnOrigValues("interface");
    $self->{_interface} = [@tmp];
    $self->{_mcast_grp} = $config->returnOrigValue("mcast-group");
    $self->{_pre_shared} = $config->returnOrigValue("pre-shared-secret");
    $self->{_keepalive_itvl} = $config->returnOrigValue("keepalive-interval");
    $self->{_dead_itvl} = $config->returnOrigValue("dead-interval");
    $self->{_monitor_dead_itvl} = $config->returnOrigValue("monitor-dead-interval");

    $config->setLevel("$level group");
    my @groups = $config->listOrigNodes();
    my $group;
    for $group (@groups) {
        my $hashref = {};
        $config->setLevel("$level group $group");
        $hashref->{_primary} = $config->returnOrigValue("primary");
        @tmp = $config->returnOrigValues("secondary");
        $hashref->{_secondary} = [@tmp];
        $hashref->{_auto_failback} = $config->returnOrigValue("auto-failback");
        @tmp = $config->returnOrigValues("monitor");
        $hashref->{_monitor} = [@tmp];
        @tmp = $config->returnOrigValues("service");
        $hashref->{_service} = [@tmp];
        $self->{_groups}->{$group} = $hashref;
    }

    return 0;
}

sub primaryNode {
    my ($self) = @_;
    my @groups = keys %{$self->{_groups}};
    my $hashref = $self->{_groups}->{$groups[0]};
    return $hashref->{_primary};
}

sub secondaryNode {
    my ($self) = @_;
    my @groups = keys %{$self->{_groups}};
    my $hashref = $self->{_groups}->{$groups[0]};
    return ${$hashref->{_secondary}}[0];
}

sub monitorNodes {
    my ($self) = @_;
    my @groups = keys %{$self->{_groups}};
    my $hashref = $self->{_groups}->{$groups[0]};
    return @{$hashref->{_monitor}};
}

sub serviceStr {
    my ($self) = @_;
    my @groups = keys %{$self->{_groups}};
    my $hashref = $self->{_groups}->{$groups[0]};
    return (join " ", @{$hashref->{_service}});
}

sub isEmpty {
    my ($self) = @_;
    return $self->{_is_empty};
}

sub authkeys {
    my ($self) = @_;
    my $key = $self->{_pre_shared};
    return (undef, "pre-shared secret not defined") if (!defined($key));
    my $str =<<EOS;
auth 1
1 sha1 $key
EOS
    return ($str, undef);
}

sub check_interfaces {
    my ($exist_only, @interfaces) = @_;
    foreach (@interfaces) {
        system("ip addr show $_ >& /dev/null");
        if ($? >> 8) {
            return "interface $_ does not exist";
        }
        next if ($exist_only);

        my $link = `ip link show $_ | grep $_`;
        my $link_test = 0;
        
        while (($link =~ /NO-CARRIER/) || !($link =~ /,UP/)) {
            $link_test++;
            sleep(1);
            $link = `ip link show $_ | grep $_`;
            if ($link_test >= 10) {
                return "interface $_ is not connected";
            }
        }
        
        system("ip addr show dev $_ |grep 'inet ' |grep -q 'scope global'");
        if ($? >> 8) {
            return "interface $_ is not configured";
        }
    }
    return undef;
}

sub ha_cf {
    my ($self) = @_;
    my @groups = keys %{$self->{_groups}};
    return (undef, "no resource group defined") if ($#groups < 0);
    return (undef, "using multiple resource groups is not supported yet")
        if ($#groups > 0);

    my $ierr = check_interfaces(0, @{$self->{_interface}});
    if (defined($ierr)) {
        return (undef, $ierr);
    }
    my $interfaces = '';
    foreach my $intf (@{$self->{_interface}}) {
        $interfaces .= "mcast $intf ";
        $interfaces .= ((defined($self->{_mcast_grp}))?  "$self->{_mcast_grp} " : "$DEFAULT_MCAST_GROUP ");
        $interfaces .= "$DEFAULT_UDP_PORT $DEFAULT_TTL 0\n";
    }

    my $kitvl = $self->{_keepalive_itvl};
    my $ditvl = $self->{_dead_itvl};
    my $mditvl = $self->{_monitor_dead_itvl};

    my $hashref = $self->{_groups}->{$groups[0]};
    my $primary = $hashref->{_primary};
    my @secondaries =  @{$hashref->{_secondary}};
    my $pings = join " ", @{$hashref->{_monitor}};
    my $auto_failback = ($hashref->{_auto_failback} eq "true") ?"on" : "off";
    my $my_name = `uname -n`;
    chomp $my_name;

    return (undef, "heartbeat interface(s) not defined") if ($interfaces eq "");
    return (undef, "keepalive interval not defined") if (!defined($kitvl));
    return (undef, "dead interval not defined") if (!defined($ditvl));
    return (undef, "monitor dead interval not defined") if (!defined($mditvl));
    return (undef, "cluster primary system not defined") if (!defined($primary));
    return (undef, "cluster secondary node(s) not defined") if ($#secondaries < 0);
    return (undef, "using multiple secondary nodes is not supported yet") if ($#secondaries > 0);
    return (undef, "dead interval must be at least $MIN_DEAD milliseconds") if ($ditvl < $MIN_DEAD);
    return (undef, "monitor dead interval must be at least $MIN_DEAD milliseconds") if ($mditvl < $MIN_DEAD);
    return (undef, "keepalive interval must be at least $MIN_KEEP milliseconds") if ($kitvl < $MIN_KEEP);
    return (undef, "dead interval must be more than twice the keepalive interval") if ($ditvl <= (2 * $kitvl));
    return (undef, "monitor dead interval must be more than twice the keepalive interval") if ($mditvl <= (2 * $kitvl));
    return (undef, "dead interval must be smaller than $DEFAULT_INITDEAD milliseconds") if ($ditvl >= $DEFAULT_INITDEAD);
    return (undef, "monitor dead interval must be smaller than $DEFAULT_INITDEAD milliseconds") if ($mditvl >= $DEFAULT_INITDEAD);
    return (undef, "the current node '$my_name' is not defined in the configuration") if (($my_name ne $primary) && ($my_name ne $secondaries[0]));

    my $monitor_str = "";
    if ($pings ne "") {
        $monitor_str = "\nping $pings\n". "respawn hacluster /usr/lib/heartbeat/ipfail";
    }

    my $wtime = int($kitvl * 2);
    if ($wtime > $ditvl) {
        $wtime = $ditvl;
    }

    # convert to seconds (HA calls "sleep" with this)
    $ditvl /= 1000;

    my $str =<<EOS;
keepalive ${kitvl}ms
deadtime ${ditvl}
warntime ${wtime}ms
initdead ${DEFAULT_INITDEAD}ms
deadping ${mditvl}ms
logfacility $DEFAULT_LOG_FACILITY
${interfaces}auto_failback $auto_failback
node $primary $secondaries[0]$monitor_str
EOS

    return ($str, undef);
}

sub isValidIPSpec {
    my $str = shift;
    my @comps = split /\//, $str;
    return 0 if ($#comps < 1);
    return 0 if ($#comps > 3);
    return 0 if (!isValidIPv4($comps[0]));

    # check optional prefix len
    if (defined($comps[1])) {
        return 0 if (!($comps[1] =~ m/^\d+$/));
        return 0 if (($comps[1] > 32) || ($comps[1] < 0));
    }

    # check optional interface
    if (defined($comps[2])) {
        return 0 if (defined(check_interfaces(1, $comps[2])));
    }

    # check optional broadcast addr
    if (defined($comps[3])) {
        return 0 if (!isValidIPv4($comps[3]));
    }
    return 1;
}

sub isValidIPv4 {
    my $str = shift;
    return 0 if (!($str =~ m/^(\d+)\.(\d+)\.(\d+)\.(\d+)$/));
    my ($a1, $a2, $a3, $a4) = ($1, $2, $3, $4);
    return 1 if (($a1 >= 0 && $a1 < 256)
        && ($a2 >= 0 && $a2 < 256)
        && ($a3 >= 0 && $a3 < 256)
        && ($a4 >= 0 && $a4 < 256));
    return 0;
}

my @service_list = ();

sub isValidService {
    my $service = shift;
    if ($service =~ /^([^:]+)::/) {
        my $script = $1;
        return 0 if (!-e "$RESOURCE_SCRIPT_DIR/$script");
        return 1;
    }
    if (scalar(@service_list) == 0) {
        opendir(SDIR, "$SERVICE_DIR")
            or (print STDERR "Error: can't open $SERVICE_DIR" && return 0);
        @service_list = grep !/^\./, readdir SDIR;
    }
    return 1 if (grep /^$service$/, @service_list);
    return 0;
}

sub haresources {
    my ($self, $conntrackd_service) = @_;
    my @groups = keys %{$self->{_groups}};
    return (undef, "no resource group defined") if ($#groups < 0);
    return (undef, "using multiple resource groups is not supported yet")
        if ($#groups > 0);

    my $hashref = $self->{_groups}->{$groups[0]};
    my $primary = $hashref->{_primary};

    my @init_services = ();
    my @ip_addresses = ();
    foreach (@{$hashref->{_service}}) {
        if (!isValidIPSpec($_)) {
            if (isValidService($_)) {
                if ($_ eq 'ipsec') {

                    # check if ipsec is configured
                    my $config = new Vyatta::Config;
                    $config->setLevel('vpn');
                    my @nodes = $config->listOrigPlusComNodes();
                    if (!grep(/^ipsec$/, @nodes)) {
                        return (undef, "ipsec is not configured");
                    }
                }
                push @init_services, $_;
            } else {
                return (undef, "\"$_\" is not a valid IP address ". "(with subnet mask length) or service name");
            }
        } else {
            push @ip_addresses, "IPaddr2-vyatta::$_";
        }
    }

    # this forces all ip addresses to be before all services, which may not
    # be the desirable behavior in all cases.
    my $ip_str = join "\\\n ", @ip_addresses;
    my $serv_str = join "\\\n ", @init_services;
    my $services = join "\\\n ", ($ip_str, $serv_str);
    return (undef, "cluster primary system not defined") if (!defined($primary));
    return (undef, "cluster service(s) not defined") if ($services eq "");

    # check if conntrack-sync is configured to use clustering
    my $config = new Vyatta::Config;
    $config->setLevel('service conntrack-sync failover-mechanism');
    my @nodes = $config->listOrigPlusComNodes();
    if (grep(/^cluster$/, @nodes)) {
        $conntrackd_service = "vyatta-cluster-conntracksync";
    }
    $services = join " ", ($services, "$conntrackd_service") if defined $conntrackd_service;
    my $str =<<EOS;
$primary $services
EOS

    return ($str, undef, @init_services);
}

sub print_str {
    my ($self) = @_;
    my $str = "cluster";
    $str .= "\n  interface " . (join ",", @{$self->{_interface}});
    $str .= "\n  pre-shared-secret $self->{_pre_shared}";
    $str .= "\n  keepalive-interval $self->{_keepalive_itvl}";
    $str .= "\n  dead-interval $self->{_dead_itvl}";
    $str .= "\n  monitor-dead-interval $self->{_monitor_dead_itvl}";
    my $group;

    foreach $group (keys %{$self->{_groups}}) {
        $str .= "\n  group $group";
        my $hashref = $self->{_groups}->{$group};
        $str .= "\n    primary $hashref->{_primary}";
        $str .= "\n    secondary " . (join ",", @{$hashref->{_secondary}});
        $str .= "\n    monitor " . (join ",", @{$hashref->{_monitor}});
        $str .= "\n    service " . (join ",", @{$hashref->{_service}});
    }

    return $str;
}

1;
